source/postcard-schema/src/max_size.rs

```diff
@@ -0,0 +1,199 @@
+//! .
+
+use crate::{
+    const_helpers::nty_eq, schema::{DataModelType, DataModelVariant, NamedType}, Schema
+};
+
+/// Calculate the max size of a type that impls Schema
+///
+/// This number must NOT be relied on for safety purposes
+/// (such as unchecked access), as manual schema impls can
+/// be wrong.
+pub const fn max_size<T: Schema>() -> Option<usize> {
+    if let Some(sz) = T::MANUAL_MAX_SIZE {
+        Some(sz)
+    } else {
+        max_size_nt(T::SCHEMA)
+    }
+}
+
+/// Calculate the max size of a NamedType
+pub const fn max_size_nt(nt: &NamedType) -> Option<usize> {
+    max_size_dmt(nt.ty)
+}
+
+/// .
+pub const fn bounded_seq_max<Outer: Schema, Inner: Schema, const N: usize>() -> Option<usize> {
+    let DataModelType::Seq(nt) = Outer::SCHEMA.ty else {
+        panic!("Not a bounded seq!");
+    };
+    assert!(nty_eq(nt, Inner::SCHEMA), "Mismatched Outer/Inner types!");
+    let size_one = max_size::<Inner>();
```

Comment:
Is this dance required to use `Inner::MANUAL_MAX_SIZE` if it's set (as opposed to using `max_size_nt()`)?

Somewhat relatedly, I'm wondering if it would make sense to embed known max sizes in the schema types somewhere in order for `postcard-dyn` to be able to take them into account. Currently, from my understanding, a client using `postcard-dyn` couldn't be told the max size of a `heapless::Vec<T, N>` because the only information in the serialized schema is the unbounded `Seq(T::SCHEMA)`. If the serialized schema included the max length, then that length could be used to size buffers even without knowing the concrete type.

Concretely, this could take the form of bundling an overall max size with an `OwnedNameType`:
```rust
struct OwnedSchema {
    ty: OwnedNamedType,
    max_size: Option<usize>,
}
```
Or it could be more tightly coupled with the schema types themselves, something like:
```rust
struct NamedType {
    name: &'static str,
    ty: &'static DataModelType,
    max_size: Option<usize>,
}
// In this case, Schema could remain as just a single associated const
trait Schema {
    const SCHEMA: &'static schema::NamedType;
}
```

Reply:
Or, less flexibly, unbounded data model types (strings, seqs, maps, etc.) could have an optional max length field. Then the max size could always be determined based on the schema. Though that might not play as nicely with schema hashing—are two schemas interchangeable if they have different max lengths for a contained string?

Reply:
> Or, less flexibly, unbounded data model types (strings, seqs, maps, etc.) could have an optional max length field. Then the max size could always be determined based on the schema. Though that might not play as nicely with schema hashing—are two schemas interchangeable if they have different max lengths for a contained string?

I'm starting to think this might be the best option:
1. It's harder to implement in subtly wrong ways (e.g., by assuming integers take the same number of bytes as their in-memory representations, which I've seen in the wild)
2. It conveys additional information (bounds on the length of collections) that could be calculated based on the max size, but not as obviously
3. If max size is needed in a `postcard-dyn`/`postcard-rpc` context, it's more compact to send `N` on the wire than `N * max_size(T)`

@jamesmunns can you think of any cases where max size could not be determined this way?